File: tools/ExternalEvaluation/CSharp/ExternalEvaluation.Service/StreamChannel.cs
#region License Information
/* HeuristicLab
 * Copyright (C) 2002-2010 Heuristic and Evolutionary Algorithms Laboratory (HEAL)
 *
 * This file is part of HeuristicLab.
 *
 * HeuristicLab is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * HeuristicLab is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with HeuristicLab. If not, see <http://www.gnu.org/licenses/>.
 */
#endregion

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using Google.ProtocolBuffers;

namespace HeuristicLab.Problems.ExternalEvaluation.Service {
  public class StreamChannel : Channel {
    private object inputLock = new object();
    private Stream input;
    private object outputLock = new object();
    private Stream output;

    public StreamChannel(Stream input, Stream output)
      : base() {
      this.input = input;
      this.output = output;
    }

    public override void Open() {
      base.Open();
    }

    public override void Send(IMessage msg) {
      lock (outputLock) {
        msg.WriteDelimitedTo(output);
        output.Flush(); // very important!
      }
    }

    public override IMessage Receive(IBuilder builder) {
      lock (inputLock) {
        // hacky, there doesn't seem to be an IBuilder.WeakMergeDelimitedFrom
        CodedInputStream cIn = CodedInputStream.CreateInstance(input);
        uint length = cIn.ReadRawVarint32();
        byte[] message = cIn.ReadRawBytes((int)length);
        builder.WeakMergeFrom(ByteString.CopyFrom(message));
        if (builder.IsInitialized)
          return builder.WeakBuild();
        else throw new EndOfStreamException("EOF reached, but message is incomplete.");
      }
    }

    public override void Close() {
      base.Close();
      input.Close();
      output.Close();
    }
  }
}

File: tools/ExternalEvaluation/Java/ExternalEvaluation.Service/src/com/heuristiclab/problems/externalevaluation/StreamChannel.java
package com.heuristiclab.problems.externalevaluation;

import java.io.*;
import com.google.protobuf.*;

public class StreamChannel extends Channel {
	private InputStream input;
	private OutputStream output;
	
	public StreamChannel(InputStream input, OutputStream output) {
		super();
		this.input = input;
		this.output = output;
	}

	@Override
	public void open() throws IOException {
		super.open();
	}
	
	@Override
	public void send(Message msg)
		throws IOException {
		synchronized (output) {
			msg.writeDelimitedTo(output);
	        output.flush(); // very important!
	    }
	}
	
	@Override
	public Message receive(Message.Builder builder)
		throws IOException {
		synchronized(input) {
			builder.mergeDelimitedFrom(input);
			if (builder.isInitialized())
				return builder.build();
			else throw new EOFException("EOF reached, but message is incomplete.");
		}
	}
	
	@Override
	public void close() throws IOException {
		super.close();
		input.close();
		output.close();
	}
}

